### PR TITLE
[10.0][ADD] sale_contract_tax_subscription

### PR DESCRIPTION
**Pull request by i-vyshnevska**
Module allows set manually taxes on the contract lines so next recurring invoice will apply this/theses taxes instead.

build is red cos enterprise  dependencies it's in WIP
troubleshooting on importing git in python

### CODE REVIEW

File: sale_contract_tax_subscription/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Camptocamp SA
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    'name': 'Sale Contract Tax Management',
+    'description': 'Allow taxes mngm directly in sale contract',
+    'author': 'camptocamp',
+    'website': 'https://camptocamp.com',
```
Comment (simahawk) [inline sale_contract_tax_subscription/__manifest__.py]: https://github.com/camptocamp/odoo-enterprise-addons

File: sale_contract_tax_subscription/data/tax.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+
+        <record id="res_company_second" model="res.company">
+            <field name="currency_id" ref="base.EUR"/>
+            <field name="partner_id" ref="base.res_partner_2"/>
```
Comment (simahawk) [inline sale_contract_tax_subscription/data/tax.xml]: :fire: demo data loaded in non demo data

File: sale_contract_tax_subscription/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+This module allows set manually taxes on the contract lines so next recurring invoice will apply
```
Comment (simahawk) [inline sale_contract_tax_subscription/readme/DESCRIPTION.rst]: you should mention it's for enterprise

File: sale_contract_tax_subscription/tests/test_tax_subscription.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        })
+
+        line_3 = line_model.create({
+            'name': 'Subscription',
+            'product_id': product_2.id,
+            'uom_id': uom.id,
+            'price_unit': 300,
+        })
+
+        self.contract.write({
```
Comment (simahawk) [inline sale_contract_tax_subscription/tests/test_tax_subscription.py]: can't you simply pass the contract id at line creation?

File: sale_contract_tax_subscription/tests/test_tax_subscription.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            'recurring_invoice_line_ids': [(6, 0, [
+                line_1.id, line_2.id, line_3.id
+            ])]
+        })
+
+    def test_tax_applying(self):
+
+        inv_id = self.contract.recurring_invoice()['domain'][0][2]
+        invoice = self.env['account.invoice'].search([('id', '=', inv_id)])
```
Comment (simahawk) [inline sale_contract_tax_subscription/tests/test_tax_subscription.py]: you have the id just browse, do not search

File: sale_contract_tax_subscription/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    'depends': [
+        'sale_contract',
+    ],
+    'data': [
+        'views/sale_subscription.xml',
+        'data/tax.xml',
```
Comment (simahawk) [inline sale_contract_tax_subscription/__manifest__.py]: I guess this is demo

File: sale_contract_tax_subscription/tests/test_tax_subscription.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    def setUp(self):
+        super(TestTaxSubscription, self).setUp()
+
+        line_model = self.env['sale.subscription.line']
+        uom = self.env.ref('product.product_uom_hour')
+        tax_15 = self.env.ref('sale_contract_tax_subscription.sale_tax_15')
+        tax_25 = self.env.ref('sale_contract_tax_subscription.sale_tax_25')
+        product_1 = self.env.ref('product.product_product_1')
+        product_2 = self.env.ref('product.product_product_2')
+        product_2.write({'taxes_id': [(6, 0, [tax_15.id])]})
```
Comment (simahawk) [inline sale_contract_tax_subscription/tests/test_tax_subscription.py]: tracking_disable

File: sale_contract_tax_subscription/models/sale_subscription.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            if tax:
+                inv_line['invoice_line_tax_ids'] = [(6, 0, tax.ids)]
+
+        return inv_line
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "sale.subscription.line"
+
+    tax_ids = fields.Many2many(
```
Comment (simahawk) [inline sale_contract_tax_subscription/models/sale_subscription.py]: shouldn't be named "additional_tax_ids"?

File: sale_contract_tax_subscription/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Camptocamp SA
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    'name': 'Sale Contract Tax Management',
+    'description': 'Allow taxes mngm directly in sale contract',
```
Comment (simahawk) [inline sale_contract_tax_subscription/__manifest__.py]: add `license` key

File: sale_contract_tax_subscription/tests/test_tax_subscription.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        inv_id = self.contract.recurring_invoice()['domain'][0][2]
+        invoice = self.env['account.invoice'].search([('id', '=', inv_id)])
+        tax = self.env.ref('sale_contract_tax_subscription.sale_tax_15')
+        inv_lines = invoice.invoice_line_ids
+
+        self.assertEqual(invoice.amount_total, 675)
+        self.assertEqual(invoice.amount_tax, 75)
+        self.assertFalse(inv_lines[0]['invoice_line_tax_ids'])
+        self.assertEqual(inv_lines[1]['invoice_line_tax_ids'].id, tax.id)
```
Comment (simahawk) [inline sale_contract_tax_subscription/tests/test_tax_subscription.py]: shouldn't you have 2 taxes?
anyway, pls add comments on what you are testing/expecting

Comment (i-vyshnevska) [reply sale_contract_tax_subscription/tests/test_tax_subscription.py]: yes, that right, comments added

**Comment by i-vyshnevska** [comment]
@simahawk can you pass review here

File: sale_contract_tax_subscription/README.rst [hunk 11]
```diff
@@ -0,0 +1,68 @@
+============================
```
Comment (simahawk) [inline sale_contract_tax_subscription/README.rst]: warning: this file won't be updated automatically as the OCA bot does. You have to compile it manually.
Or, keep it simple and use a plaing README.rst file w/out fragments

File: sale_contract_tax_subscription/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    'author': 'camptocamp',
+    'website': 'https://camptocamp.com',
+    'depends': [
+        'sale_contract',
+    ],
+    'data': [
+        'views/sale_subscription.xml',
+        'data/tax.xml',
+    ],
+    'application': False,
```
Comment (simahawk) [inline sale_contract_tax_subscription/__manifest__.py]: you can remove this flag

File: sale_contract_tax_subscription/data/tax.xml [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+
+        <record id="res_company_second" model="res.company">
```
Comment (simahawk) [inline sale_contract_tax_subscription/data/tax.xml]: this is for? test/demo only?

File: sale_contract_tax_subscription/models/sale_subscription.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        help="Taxes to change default fiscal politics",
+        domain="[('type_tax_use','=','sale')]"
+    )
+    applied_tax_desc = fields.Char(
+        string="Applied taxes",
+        help="Taxes which will be applied to current line",
+        compute="_compute_applied_taxes",
+    )
+
+    def _compute_applied_taxes(self):
```
Comment (simahawk) [inline sale_contract_tax_subscription/models/sale_subscription.py]: api.depends missing, name should be `_compute_applied_tax_desc`

File: sale_contract_tax_subscription/models/sale_subscription.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        Compute names of applied tax to aware user which taxes would be applied
+        to line in invoice, this data applicable only to the current moment
+        here we don't care about future fiscal politics
+        """
+
+        contract = self[0].analytic_account_id
```
Comment (simahawk) [inline sale_contract_tax_subscription/models/sale_subscription.py]: hmm are we sure you'll get always the right record here?

File: sale_contract_tax_subscription/tests/test_tax_subscription.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        inv_id = self.contract.recurring_invoice()['domain'][0][2]
+        invoice = self.env['account.invoice'].browse(inv_id)
+        tax = self.env.ref('sale_contract_tax_subscription.sale_tax_15')
+        inv_lines = invoice.invoice_line_ids
+
+        self.assertEqual(invoice.amount_total, 675)
+        self.assertEqual(invoice.amount_tax, 75)
+        self.assertFalse(inv_lines[0]['invoice_line_tax_ids'])
+        # only one tax applied, other belong to other company should be cut
+        self.assertEqual(inv_lines[1]['invoice_line_tax_ids'].id, tax.id)
```
Comment (simahawk) [inline sale_contract_tax_subscription/tests/test_tax_subscription.py]: test the computation as well

File: sale_contract_tax_subscription/data/tax.xml (hunk 13 above)
Comment (i-vyshnevska) [reply sale_contract_tax_subscription/data/tax.xml]: for tests

Comment (simahawk) [reply sale_contract_tax_subscription/data/tax.xml]: then you have to load it in `demo`

File: sale_contract_tax_subscription/tests/test_tax_subscription.py (hunk 16 above)
Comment (i-vyshnevska) [reply sale_contract_tax_subscription/tests/test_tax_subscription.py]: @simahawk updated

**Pull request merged**